app/rdf/queries.ts

```diff
@@ -784,103 +789,130 @@ export const hasHierarchy = (dim: CubeDimension) => {
+        throw new Error(
+          `Dimension <${iri}> has dataType 'langString', which is not supported by Visualize. In order to fix it, change the dataType to 'string' in the cube definition.`
+        );
+      }
 
-    const dimensionHasHierarchy = hasHierarchy(cubeDimension);
-    const toRDFValue = (d: string): NamedNode | Literal => {
-      return dataType && !dimensionHasHierarchy
-        ? parsedCubeDimension.data.hasUndefinedValues &&
-          d === DIMENSION_VALUE_UNDEFINED
-          ? rdf.literal("", ns.cube.Undefined)
-          : rdf.literal(d, dataType)
-        : rdf.namedNode(d);
-    };
-
-    return filter.type === "single"
-      ? [
-          filterDimension.filter.eq(
-            toRDFValue(
-              typeof filter.value === "number"
-                ? filter.value.toString()
-                : filter.value
-            )
-          ),
-        ]
-      : filter.type === "multi"
-      ? // If values is an empty object, we filter by something that doesn't exist
-        [
-          filterDimension.filter.in(
-            Object.keys(filter.values).length > 0
-              ? Object.entries(filter.values).flatMap(([iri, selected]) =>
-                  selected ? [toRDFValue(iri)] : []
-                )
-              : [rdf.namedNode("EMPTY_VALUE")]
-          ),
-        ]
-      : filter.type === "range"
-      ? [
-          filterDimension.filter.gte(toRDFValue(filter.from)),
-          filterDimension.filter.lte(toRDFValue(filter.to)),
-        ]
-      : [];
-  });
+      const dimensionHasHierarchy = hasHierarchy(cubeDimension);
+      const toRDFValue = (d: string): NamedNode | Literal => {
+        return dataType && !dimensionHasHierarchy
+          ? parsedCubeDimension.data.hasUndefinedValues &&
+            d === DIMENSION_VALUE_UNDEFINED
+            ? rdf.literal("", ns.cube.Undefined)
+            : rdf.literal(d, dataType)
+          : rdf.namedNode(d);
+      };
+
+      switch (filter.type) {
+        case "single": {
+          if (isDynamicMaxValue(filter.value)) {
+            if (cubeDimension.maxInclusive) {
+              return [
+                filterDimension.filter.eq(
+                  toRDFValue(cubeDimension.maxInclusive.value)
+                ),
+              ];
+            }
+
+            // Ideally we would query the max value directly in the observations
+            // query, but it doesn't seem to be supported by the cube-view-query
+            const maxValueQuery = SELECT`(MAX(?value) as ?value)`
+              .WHERE` <${cube.term?.value}> <https://cube.link/observationSet> ?observationSet .
+  ?observationSet <https://cube.link/observation> ?source .
+  ?source <${cubeDimension.path?.value}> ?value .`;
+            const maxValueRaw = await maxValueQuery.execute(
+              sparqlClient.query,
+              { operation: "postUrlencoded" }
+            );
+            const maxValue = maxValueRaw?.[0]?.value?.value as string;
```

Comment:
Function could be extracted for clarity ? It could aso be memoized ?

It made me think that if we'd use react-query, we could cache the sparql client results. Then I remember that we are using urql for it's graphql abilities 🤷

Reply:
Yup, it can be extracted 👍 But for memoization I am not sure, how would you memoize it, I guess when it's extracted it's already "memoized"? (as we are not in the useMemo world here in the backend 😅)